### PR TITLE
KC 1.7 combining app and realm roles

### PR DESCRIPTION
**Pull request by hrstoyanov**
Max,
here is another PR that combines roles.

### CODE REVIEW

File: errai-security/errai-security-keycloak/src/main/java/org/jboss/errai/security/keycloak/KeycloakAuthenticationService.java [hunk 1]
```diff
@@ -230,7 +230,17 @@ protected User createKeycloakUser(final AccessToken accessToken) {
   }
 
   private Collection<? extends Role> createRoles(final AccessToken accessToken) {
-    Set<String> roleNames = accessToken.getResourceAccess(accessToken.getIssuedFor()).getRoles();
+    Set<String roleNames = new HashSet<>();
```
Comment (mbarkley) [inline errai-security/errai-security-keycloak/src/main/java/org/jboss/errai/security/keycloak/KeycloakAuthenticationService.java]: There are a few issues with this line:
* `Set<String` is missing a closing angle bracket.
* Errai 3.x targets Java 1.6 so we can't use the `<>` operator. It should be `<String>` instead.
* This is the first usage of `HashSet` in this file, so there's a missing import for `java.util.HashSet`.

File: errai-security/errai-security-keycloak/src/main/java/org/jboss/errai/security/keycloak/KeycloakAuthenticationService.java [hunk 2]
```diff
@@ -230,7 +230,17 @@ protected User createKeycloakUser(final AccessToken accessToken) {
...
+    //Add app roles first, if any
+    AccessToken.Access access = accessToken.getResourceAccess(accessToken.getIssuedFor());
+    if(access!=null && access.getRoles()!=null){
```
Comment (mbarkley) [inline errai-security/errai-security-keycloak/src/main/java/org/jboss/errai/security/keycloak/KeycloakAuthenticationService.java]: To be consistent with the project's code formatting, this line should be formatted like so:
`if (access != null && access.getRoles() != null) {`

File: errai-security/errai-security-keycloak/src/main/java/org/jboss/errai/security/keycloak/KeycloakAuthenticationService.java [hunk 3]
```diff
@@ -230,7 +230,17 @@ protected User createKeycloakUser(final AccessToken accessToken) {
...
+        roleNames.addAll(access.getRoles());
```
Comment (mbarkley) [inline errai-security/errai-security-keycloak/src/main/java/org/jboss/errai/security/keycloak/KeycloakAuthenticationService.java]: To be consistent with the project's code formatting, this line should use 2 spaces for indentation.

File: errai-security/errai-security-keycloak/src/main/java/org/jboss/errai/security/keycloak/KeycloakAuthenticationService.java [hunk 4]
```diff
@@ -230,7 +230,17 @@ protected User createKeycloakUser(final AccessToken accessToken) {
...
+    }
+    //Add realm roles next, if any
+    AccessToken.Access realmAccess = accessToken.getRealmAccess();
+    if(realmAccess!=null && realmAccess.getRoles()!=null){
```
Comment (mbarkley) [inline errai-security/errai-security-keycloak/src/main/java/org/jboss/errai/security/keycloak/KeycloakAuthenticationService.java]: This line should be reformatted similarly to line 236.

File: errai-security/errai-security-keycloak/src/main/java/org/jboss/errai/security/keycloak/KeycloakAuthenticationService.java [hunk 5]
```diff
@@ -230,7 +230,17 @@ protected User createKeycloakUser(final AccessToken accessToken) {
...
+        roleNames.addAll(realmAccess.getRoles());
```
Comment (mbarkley) [inline errai-security/errai-security-keycloak/src/main/java/org/jboss/errai/security/keycloak/KeycloakAuthenticationService.java]: Please see the line 237 comment.

**Comment by mbarkley** [comment]
Hi Hristo,

The logic here looks good. But I've added some in-line comments regarding some compilation and formatting issues that need to be addressed.

If you can fix those issues, I'd be happy to merge this. Also, if you could squash all of your changes into a single commit, that would be appreciated.

Cheers.

**Comment by hrstoyanov** [comment]
OK .. My ide had issues with maven, so it fail to catch the issues.

**Comment by hrstoyanov** [comment]
Max,
I did the changes, but made a mess of my git branch and squashing

**Comment by hrstoyanov** [comment]
Max, take a look. I think I failed to squash the commits though :-(

**Comment by hrstoyanov** [comment]
... So if you want, ignore my PRs, get the attached file and put the changes in.

**Comment by hrstoyanov** [comment]
,

**Comment by mbarkley** [comment]
Hi Hristo,

I've squashed your commits [here](https://github.com/errai/errai/commit/7d6159de719cf49efe20be0bd8443a7ff0947a2e) on 3.2 and [here](https://github.com/errai/errai/commit/aa778a8a167f5a6deeb3c493afeedd6723b227d5) on master.

I've also deployed artifacts fro 3.2.3-SNAPSHOT.

Cheers.

**Comment by hrstoyanov** [comment]
Thanks Max! I will test it off the official maven repo tonight. I expect the KC guys to come out with 1.8 any day now, at which point which I will test it and then submit another minor PR. I see KC now has "groups" in addition to "roles", but have not seen them in KC 1.7 access token API or admin console. If I see them in KC 1.8, there will be more PRs, as it seems like Errai has groups too. 

A Suggestion: PicketLink is officially dead for an year now, it might be a good idea to drop it from Errai 4 and demo, unless you ar aware of someone using it (it can still stay Errai 3). The current demo is confusing for newcomers as it tries to show off both PicketLink and KeyCloak usage. At a minimum, perhaps splitting it into 2 separate demos might help.

**Pull request closed without merging**